apps/src/code-studio/components/progress/UnitCalendar.jsx

```diff
@@ -0,0 +1,197 @@
+  weekColumn: {
+    minWidth: 100,
+    backgroundColor: color.purple,
+    color: 'white',
+    textAlign: 'center',
+    border: '1px solid ' + color.purple,
+    borderCollapse: 'collapse',
+    fontWeight: 'bold',
+    minHeight: 50
+  },
+  scheduleColumn: {
+    border: '1px solid ' + color.purple,
+    borderCollapse: 'collapse',
+    width: weekWidth,
+    display: 'flex',
+    minHeight: 50
+  },
+  table: {
+    border: '1px solid ' + color.purple,
+    borderCollapse: 'collapse',
+    width: '100%'
+  },
+  key: {
+    border: '1px solid ' + color.purple,
+    borderCollapse: 'collapse',
+    width: '100%',
+    marginTop: 20
+  },
+  keyIcon: {
+    marginRight: 5
+  },
+  keySection: {
+    display: 'flex',
+    justifyContent: 'space-around',
+    width: '90%',
+    alignItems: 'center',
+    fontSize: 15
+  }
+};
+
+export default class UnitCalendar extends React.Component {
+  static propTypes = {
+    weeklyInstructionalMinutes: PropTypes.number.isRequired,
+    lessons: PropTypes.arrayOf(unitCalendarLesson).isRequired
+  };
+  constructor(props) {
+    super(props);
+    this.state = {
+      hovering: ''
+    };
+  }
+  generateSchedule = () => {
+    const {lessons, weeklyInstructionalMinutes} = this.props;
+    const lessonsCopy = _.cloneDeep(lessons);
+    let allWeeks = [];
+    let currWeek = [];
+    let currMinutes = 0;
+
+    // splitting lessons across weeks
+    let prev = null;
+    lessonsCopy.forEach(lesson => {
+      lesson.isStart = false;
+      lesson.isEnd = false;
+      lesson.isMajority = false;
+      let lessonDuration = lesson.duration;
+      while (lessonDuration > 0) {
+        let thisChunk = null;
+        if (currMinutes + lessonDuration < weeklyInstructionalMinutes) {
+          // If the rest of the current lesson fits into this week, put it in the schedule.
+          currMinutes = currMinutes + lessonDuration;
+          let thisLesson = _.cloneDeep(lesson);
+          thisLesson.isEnd = true;
+          currWeek.push(thisLesson);
+          thisChunk = thisLesson;
+          lessonDuration = 0;
+        } else if (currMinutes < weeklyInstructionalMinutes - 5) {
+          // If there's more than 5 minutes left in the week,
+          // add as much of the lesson as you can to this week.
+          let partialLesson = _.cloneDeep(lesson);
+          partialLesson.duration = weeklyInstructionalMinutes - currMinutes;
+          currWeek.push(partialLesson);
+          thisChunk = partialLesson;
+          lessonDuration = lessonDuration - partialLesson.duration;
+          lesson.duration = lessonDuration;
+          currMinutes = weeklyInstructionalMinutes;
+        } else {
+          // If there isn't enough time in this week to add this lesson, start a new week.
+          allWeeks.push(currWeek);
+          currWeek = [];
+          currMinutes = 0;
+        }
+        if (thisChunk) {
+          if (!prev || prev.title !== thisChunk.title) {
+            thisChunk.isStart = true;
```

Comment:
Could you tell if a section is the start based on when you are about to add it by checking if lessonDuration is equal to lesson.duration?

Reply:
you're right, I really just need to do the majority check at the end.

Reply:
I think there might be a way to figure out majority as you go but lets move start and then take a look. This is a a tricky function and you have done a great job with it. I was just trying to find ways to make it a little bit easier to maintain going forward if needed

Reply:
I can move the majority check to inside the if/else if? I'm not sure how else to track it